thread/common/threadhelpers.cpp

```diff
@@ -195,6 +195,246 @@ omrthread_spinlock_swapState(omrthread_monitor_t monitor, uintptr_t newState)
+	if (oldState == VM_AtomicSupport::lockCompareExchange(
+			(volatile uintptr_t *)&monitor->queueTail,
+			(uintptr_t)oldState,
+			(uintptr_t)mcsNode)
+	) {
+		/* monitor->spinlockState is maintained for compatibility with the existing omrthread API. */
+		monitor->spinlockState = J9THREAD_MONITOR_SPINLOCK_OWNED;
+
+		/* Each omrthread_t maintains a stack of MCS nodes in case it acquires multiple locks incrementally.
+		 * After the lock is acquired, the MCS node is pushed to the thread's MCS node stack. Assumption:
+		 * the locks are acquired and released in order. Example:
+		 *     acquire(lock1)
+		 *         acquire(lock2)
+		 *         release(lock2)
+		 *    release(lock1)
+		 *
+		 * The following is not allowed:
+		 *     acquire(lock1)
+		 *         acquire(lock2)
+		 *    release(lock1)
+		 *         release(lock2)
+		 */
+		if (NULL == self->mcsNodes->stackHead) {
+			self->mcsNodes->stackHead = mcsNode;
+			mcsNode->stackNext = NULL;
+		} else {
+			omrthread_mcs_node_t oldMCSNode = self->mcsNodes->stackHead;
+			self->mcsNodes->stackHead = mcsNode;
+			mcsNode->stackNext = oldMCSNode;
+		}
+		result = 0;
+	}
+
+	return result;
+}
+
+/**
+ * Unlock the MCS lock.
+ *
+ * @param[in] self the current omrthread_t
+ * @param[in] monitor the monitor to be released
+ *
+ * @return void
+ */
+void
+omrthread_mcs_unlock(omrthread_t self, omrthread_monitor_t monitor)
+{
+	/* The MCS node at the top of the stack corresponds to the current monitor.
+	 * Refer to the assumption stated in omrthread_mcs_lock and omrthread_mcs_trylock.
+	 */
+	omrthread_mcs_node_t mcsNode = self->mcsNodes->stackHead;
+	ASSERT(mcsNode != NULL);
+
+	/* Get the successor of the mcsNode. */
+	omrthread_mcs_node_t successor = (omrthread_mcs_node_t)VM_AtomicSupport::add((volatile uintptr_t *)&mcsNode->queueNext, (uintptr_t)0);
+	if (NULL == successor) {
+		/* If no successor exists, then mcsNode is at the tail of the MCS lock queue. Release
+		 * the lock by replacing the mcsNode at the tail of the queue with NULL.
+		 */
+		uintptr_t newState = 0;
+		if ((uintptr_t)mcsNode == VM_AtomicSupport::lockCompareExchange((volatile uintptr_t *)&monitor->queueTail, (uintptr_t)mcsNode, newState)) {
+			monitor->spinlockState = J9THREAD_MONITOR_SPINLOCK_UNOWNED;
+			goto lockReleased;
+		}
+
+		/* Another thread is recording a successor in mcsNode->queueNext. Wait for the thread
+		 * to record the successor.
+		 */
+		while (NULL == (successor = (omrthread_mcs_node_t)VM_AtomicSupport::add((volatile uintptr_t *)&mcsNode->queueNext, (uintptr_t)0)));
+	}
+	/* monitor->spinlockState is maintained for compatibility with the existing omrthread API. */
+	monitor->spinlockState = J9THREAD_MONITOR_SPINLOCK_UNOWNED;
+
+	/* Allow the successor to acquire the lock by resetting its blocked field. */
+	successor->blocked = 0;
+
+lockReleased:
+	/* Pop the mcsNode from the thread's MCS node stack. */
+	self->mcsNodes->stackHead = mcsNode->stackNext;
```

Comment:
If the lock acquire code (Line 281-288) is run right after `successor->blocked` is set to 0, and before this current line, does it break the following unlock operations after this one?

Reply:
Every thread has its own `J9Thread->mcsNodes`, which contains the `stackHead` and `J9Pool` of MCS nodes. Each thread only reads/writes its own `J9Thread->mcsNodes`.

The thread, which will execute the lock acquire code after `successor->blocked` is set to 0, will access its own `J9Thread->mcsNodes`. No synchronization is required to access `J9Thread->mcsNodes`. Hence, the unlock operation should work fine.